serverless/internal/storage/webstorage/fs.go

```
@@ -0,0 +1,290 @@
+// limitations under the License.
+
+// +build wasm
+
+// Package webstorage provides a simple log storage implementation based on webstorage.
+// It only really makes sense for wasm targets.
+package webstorage
+
+import (
+	"crypto/sha256"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"syscall/js"
+
+	"github.com/golang/glog"
+	"github.com/google/trillian-examples/formats/log"
+	"github.com/google/trillian-examples/serverless/api"
+	"github.com/google/trillian-examples/serverless/api/layout"
+	"github.com/google/trillian-examples/serverless/internal/storage"
+)
+
+// Storage is a serverless storage implementation which uses webstorage entries to store tree state.
+// The storage key format is:
+//  <root>/leaves/aa/bb/cc/ddeeff...
+//  <root>/leaves/pending/aabbccddeeff...
+//  <root>/seq/aa/bb/cc/ddeeff...
+//  <root>/tile/<level>/aa/bb/ccddee...
+//  <root>/checkpoint
+//
+// The functions on this struct are not thread-safe.
+type Storage struct {
+	// root is the prefix of all webstorage keys which store tree data.
+	root string
+	// nextSeq is a hint to the Sequence func as to what the next available
+	// sequence number is to help performance.
+	// Note that nextSeq may be <= than the actual next available number, but
+	// never greater.
+	nextSeq uint64
+	// checkpoint is the latest known checkpoint of the log.
+	checkpoint log.Checkpoint
+}
+
+const leavesPendingPathFmt = "leaves/pending/%0x"
+
+func getStorage() js.Value {
+	return js.Global().Get("sessionStorage")
+}
+
+func exists(k string) bool {
+	_, err := get(k)
+	return err == nil
+}
+
+func get(k string) ([]byte, error) {
+	v := getStorage().Call("getItem", js.ValueOf(k))
+	if js.Null().Equal(v) {
+		return nil, os.ErrNotExist
+	}
+	return []byte(v.String()), nil
+}
+
+func set(k string, v []byte) error {
+	getStorage().Call("setItem", js.ValueOf(k), js.ValueOf(string(v)))
+	// TODO(al) read back and check
```

Is this a best practice thing, or literally something you observed while playing with this?

---

This is because I have no idea how these calls are supposed to let the Go caller know when if it's failed to do its thing.
In the JS world it should raise an exception but I can't find any docs on what that looks like from Go's PoV...
`Get()` will at least return a `Null` if the thing I asked for isn't there.